File: app/src/main/java/com/marceljurtz/lifecounter/models/Game.java
package com.marceljurtz.lifecounter.models;

import android.content.SharedPreferences;

import com.marceljurtz.lifecounter.enums.ClickTypeEnum;
import com.marceljurtz.lifecounter.enums.OperatorEnum;
import com.marceljurtz.lifecounter.enums.PlayerIdEnum;

public class Game {

    private SharedPreferences preferences;

    private Player[] players;

    public Game(SharedPreferences preferences, int playerAmount) {
        this.preferences = preferences;
        this.players = new Player[playerAmount];
    }

    public void saveGameState(SharedPreferences preferences) {
        if(players.length == 4) {
            PreferenceManager.save4PlayerPointsData(preferences, players);
        } else {
            PreferenceManager.save2PlayerPointsData(preferences, players);
        }
    }

    public void loadGameState(SharedPreferences preferences, int playeramount) {
        if(playeramount == 4) {
            players = PreferenceManager.load4PlayerPointsData(preferences);
        } else {
            players = PreferenceManager.load2PlayerPointsData(preferences);
        }
    }

    public Player[] getPlayers() {
        return players;
    }

    public void updateLifepoints(PlayerIdEnum playerIdEnum, ClickTypeEnum clickTypeEnum, OperatorEnum operatorEnum) {

        int amount = PreferenceManager.getDefaultShortclickPoints();
        if(clickTypeEnum.equals(ClickTypeEnum.LONG)) amount = PreferenceManager.getLongclickPoints(preferences);

        if(operatorEnum.equals(OperatorEnum.SUBSTRACT)) amount *= -1;

        switch(playerIdEnum) {
            case ONE:
                players[0].updateLifepoints(amount);
                break;
            case TWO:
                players[1].updateLifepoints(amount);
                break;
            case THREE:
                players[2].updateLifepoints(amount);
                break;
            case FOUR:
                players[3].updateLifepoints(amount);
                break;
            default:
                // Nothing to do
                break;
        }
    }

    public void resetLifePoints() {
        for(Player player : players) {
            player.resetPoints(preferences);
            player.updatePoisonpoints(0);
        }
    }

    public void updatePoisonpoints(PlayerIdEnum playerIdEnum, ClickTypeEnum clickTypeEnum, OperatorEnum operatorEnum) {

        int amount = PreferenceManager.getDefaultShortclickPoints();
        if(clickTypeEnum.equals(ClickTypeEnum.LONG)) amount = PreferenceManager.getLongclickPoints(preferences);

        if(operatorEnum.equals(OperatorEnum.SUBSTRACT)) amount *= -1;

        switch(playerIdEnum) {
            case ONE:
                players[0].updatePoisonpoints(amount);
                break;
            case TWO:
                players[1].updatePoisonpoints(amount);
                break;
            case THREE:
                players[2].updatePoisonpoints(amount);
                break;
            case FOUR:
                players[3].updatePoisonpoints(amount);
                break;
            default:
                // Nothing to do
                break;
        }
    }

    public int getPlayerLifepoints(PlayerIdEnum playerIdEnum) {
        switch(playerIdEnum) {
            case ONE :
                return players[0].getLifePoints();
            case TWO:
                return players[1].getLifePoints();
            case THREE:
                return players[2].getLifePoints();
            case FOUR:
                return players[3].getLifePoints();
            default:
                return 0;
        }
    }

    public int getPlayerPoisonpoints(PlayerIdEnum playerIdEnum) {
        switch(playerIdEnum) {
            case ONE :
                return players[0].getPoisonPoints();
            case TWO:
                return players[1].getPoisonPoints();
            case THREE:
                return players[2].getPoisonPoints();
            case FOUR:
                return players[3].getPoisonPoints();
            default:
                return 0;
        }
    }

    public Player getPlayer1() { return players[0]; }
    public Player getPlayer2() { return players[1]; }
    public Player getPlayer3() { return players[2]; }
    public Player getPlayer4() { return players[3]; }
}

File: app/src/main/java/com/marceljurtz/lifecounter/models/Game.kt
package com.marceljurtz.lifecounter.models

import android.content.SharedPreferences

import com.marceljurtz.lifecounter.enums.ClickTypeEnum
import com.marceljurtz.lifecounter.enums.OperatorEnum
import com.marceljurtz.lifecounter.enums.PlayerIdEnum

class Game(private val preferences: SharedPreferences, playerAmount: Int) {

    var players: Array<Player?>? = null
        private set

    val player1: Player
        get() = players!!.get(0)!!
    val player2: Player
        get() = players!!.get(1)!!
    val player3: Player
        get() = players!!.get(2)!!
    val player4: Player
        get() = players!!.get(3)!!

    init {
        this.players = arrayOfNulls(playerAmount)
    }

    fun saveGameState(preferences: SharedPreferences) {
        if (players!!.size == 4) {
            PreferenceManager.save4PlayerPointsData(preferences, players!!)
        } else {
            PreferenceManager.save2PlayerPointsData(preferences, players!!)
        }
    }

    fun loadGameState(preferences: SharedPreferences, playeramount: Int) {
        if (playeramount == 4) {
            players = PreferenceManager.load4PlayerPointsData(preferences)
        } else {
            players = PreferenceManager.load2PlayerPointsData(preferences)
        }
    }

    fun updateLifepoints(playerIdEnum: PlayerIdEnum, clickTypeEnum: ClickTypeEnum, operatorEnum: OperatorEnum) {

        var amount = PreferenceManager.defaultShortclickPoints
        if (clickTypeEnum == ClickTypeEnum.LONG) amount = PreferenceManager.getLongclickPoints(preferences)

        if (operatorEnum == OperatorEnum.SUBSTRACT) amount *= -1

        when (playerIdEnum) {
            PlayerIdEnum.ONE -> players!![0]!!.updateLifepoints(amount)
            PlayerIdEnum.TWO -> players!![1]!!.updateLifepoints(amount)
            PlayerIdEnum.THREE -> players!![2]!!.updateLifepoints(amount)
            PlayerIdEnum.FOUR -> players!![3]!!.updateLifepoints(amount)
            else -> {
            }
        }// Nothing to do
    }

    fun resetLifePoints() {
        for (player in players!!) {
            player!!.resetPoints(preferences)
            player!!.updatePoisonpoints(0)
        }
    }

    fun updatePoisonpoints(playerIdEnum: PlayerIdEnum, clickTypeEnum: ClickTypeEnum, operatorEnum: OperatorEnum) {

        var amount = PreferenceManager.defaultShortclickPoints
        if (clickTypeEnum == ClickTypeEnum.LONG) amount = PreferenceManager.getLongclickPoints(preferences)

        if (operatorEnum == OperatorEnum.SUBSTRACT) amount *= -1

        when (playerIdEnum) {
            PlayerIdEnum.ONE -> players!![0]!!.updatePoisonpoints(amount)
            PlayerIdEnum.TWO -> players!![1]!!.updatePoisonpoints(amount)
            PlayerIdEnum.THREE -> players!![2]!!.updatePoisonpoints(amount)
            PlayerIdEnum.FOUR -> players!![3]!!.updatePoisonpoints(amount)
            else -> {
            }
        }// Nothing to do
    }

    fun getPlayerLifepoints(playerIdEnum: PlayerIdEnum): Int {
        when (playerIdEnum) {
            PlayerIdEnum.ONE -> return players!![0]!!.lifePoints
            PlayerIdEnum.TWO -> return players!![1]!!.lifePoints
            PlayerIdEnum.THREE -> return players!![2]!!.lifePoints
            PlayerIdEnum.FOUR -> return players!![3]!!.lifePoints
            else -> return 0
        }
    }

    fun getPlayerPoisonpoints(playerIdEnum: PlayerIdEnum): Int {
        when (playerIdEnum) {
            PlayerIdEnum.ONE -> return players!![0]!!.poisonPoints
            PlayerIdEnum.TWO -> return players!![1]!!.poisonPoints
            PlayerIdEnum.THREE -> return players!![2]!!.poisonPoints
            PlayerIdEnum.FOUR -> return players!![3]!!.poisonPoints
            else -> return 0
        }
    }
}
